### ISSUE TITLE
CA1823 with perpetual timer

### ISSUE THREAD

**Issue by adraut** [dotnet/roslyn-analyzers#1906]
#### Analyzer package
Microsoft.CodeQuality.Analyzers

#### Package Version
2.6.2
Visual Studio 15.9.0

#### Diagnostic ID
CA1823

#### Repro steps
```
Public Class VbSharedTimer
Private Shared ReadOnly CleanUpTimer as New System.Threading.Timer(AddressOf CleanUp, Nothing, 1000, 1000)

Private Shared Sub CleanUp(state as Object)
  System.Threading.Thread.Sleep(1)
End Sub
End Class
```
1. Build

#### Expected behavior
Diagnostic is not reported during build.

#### Actual behavior
`warning CA1823: Unused field 'CleanUpTimer'`

This one is a bit odd in that the diagnostic warning is reported in the build log, but is not displayed in the Error List when the view is set to `Build + IntelliSense`.  It does display when the view is set to `Build`.

**Comment by adraut** [comment]
Even more odd.  I closed the solution and re-opened it.  The warning no longer occurs.  I had done multiple rebuilds prior to closing and the warning was consistently reported from the compiler, but isn't now.

**Comment by mavasani** [comment]
Expected behavior seems to be incorrect, as the field is indeed unused and so the diagnostic should be reported during build.

Actual behavior is by design if full solution analysis is not turned on, see details [here](https://docs.microsoft.com/en-us/visualstudio/code-quality/how-to-enable-and-disable-full-solution-analysis-for-managed-code?view=vs-2017). See https://github.com/dotnet/roslyn-analyzers/issues/1559#issuecomment-363196049 for discussions on a similar issue.

To summarize - you can turn full solution analysis ON if you want consistent behavior between Build and Build + Intellisense. If full solution analysis is OFF, then CA1823 will be reported for "Build" when you do an explicit build, but these will get removed from error list when you open that file and you are in  in "Build + Intellisense" mode. The latter was very recently fixed in Roslyn, so with Dev16 Preview2, opening the file will not clear such diagnostics from error list.
